lib/BitLib.sol

```diff
@@ -2,12 +2,13 @@
 pragma solidity ^0.8.17;
 
 library BitLib {
+  // Returns the number of zeroes in x that do not have a 1 to their left, eg ctz(0)=256, ctz(1)=0
   function ctz(uint x) internal pure returns (uint c) {
     assembly ("memory-safe") {
       // Isolate the least significant bit
       x := and(x, add(not(x), 1))
 
-      // Get first 3 bits of c
+      // Get first 3 bits of c, this is the unusual part
```

Comment:
There's a common trick documented in [this paper](http://supertech.csail.mit.edu/papers/debruijn.pdf) and [this page](https://graphics.stanford.edu/~seander/bithacks.html#IntegerLogDeBruijn) (also more on [stackoverflow](https://stackoverflow.com/questions/68677658/how-does-one-find-the-floor-of-the-log-base-2-of-an-n-bit-integer-using-bitwise) and [wikipedia](https://en.wikipedia.org/wiki/De_Bruijn_sequence) to help you find the position of an isolated one in a bit sequence.

Starting from solady's solidity implementation of this idea (see below in the same file), I have found a trick to be cheaper with 256 words (normally you have to find the first 3 bits of the 1 position "by hand" then you can use deBruijn for the remaining 5 bits). I haven't yet written a complete explainer because I don't think we'll end up using it, but I'll write something up in the future. I don't think we'll use it because it seems like our bitmaps will be on 64 or 128 bits, in which case the method I use here is not the cheapest gaswise.

Reply:
Ok great. If we end up keeping (a variant of this) these reference will be great to have in the code.